src/core/settings/plugins/keyfile/nms-keyfile-writer.c

```diff
@@ -416,33 +416,75 @@ _internal_write_connection(NMConnection                   *connection,
 
     /* NETPLAN: write only non-volatile files to /etc/netplan/... */
     if (!is_volatile) {
-        g_autofree gchar* ssid = g_key_file_get_string(kf_file, "wifi", "ssid", NULL);
-        g_autofree gchar* escaped_ssid = ssid ?
+        g_autofree gchar *ssid = g_key_file_get_string(kf_file, "wifi", "ssid", NULL);
+        g_autofree gchar *escaped_ssid = ssid ?
                                          g_uri_escape_string(ssid, NULL, TRUE) : NULL;
-        g_autofree gchar* netplan_id = (existing_path && strstr(existing_path, "system-connections/netplan-")) ?
-                                       netplan_get_id_from_nm_filename(existing_path, ssid) : NULL;
-        netplan_clear_netdefs();
-
+        g_autofree gchar *netplan_id = NULL;
+        g_autofree gchar *actual_netplan_id = NULL;
+        ssize_t netplan_id_size = 0;
+        NetplanNetDefinition *netdef = NULL;
+        NetplanParser *npp = NULL;
+        NetplanState *np_state = NULL;
+        NetplanStateIterator state_iter;
         const gchar* kf_path = path;
+
+        if (existing_path && strstr(existing_path, "system-connections/netplan-")) {
+            netplan_id = g_malloc0(strlen(existing_path));
+            netplan_id_size = netplan_get_id_from_nm_filepath(existing_path, ssid, netplan_id, strlen(existing_path) - 1);
+            if (netplan_id_size <= 0) {
+                g_free(netplan_id);
+                netplan_id = NULL;
+            }
+        }
+
         if (netplan_id && existing_path) {
             GFile* from = g_file_new_for_path(path);
             GFile* to = g_file_new_for_path(existing_path);
             g_file_copy(from, to, G_FILE_COPY_OVERWRITE, NULL, NULL, NULL, NULL);
             kf_path = existing_path;
         }
+
         // push keyfile into libnetplan for parsing (using existing_path, if available,
         // to be able to extract the original netdef_id and override existing settings)
-        if (!netplan_parse_keyfile(kf_path, &local_err)) {
+        npp = netplan_parser_new();
+
+        if (!netplan_parser_load_keyfile(npp, kf_path, &local_err)) {
             g_set_error (error, NM_SETTINGS_ERROR, NM_SETTINGS_ERROR_FAILED,
                          "netplan: YAML translation failed");
```

Comment:
We should append the `local_err->message` string here, to forward libnetplan's error message.